framework/Util/Cap.swift

```diff
@@ -0,0 +1,12 @@
+import Foundation
+public func cap<T :Numeric & Comparable>(value v : T, min a : T, max b : T) -> T {
```

Comment:
I've commonly seen this function named `clamp` in GLSL and other libraries. Here's a possibly simpler definition of that:

https://stackoverflow.com/a/46799935/19679

```
public func clamp<T>(_ value: T, minValue: T, maxValue: T) -> T where T : Comparable {
    return min(max(value, minValue), maxValue)
}
```

Reply:
Oh yes, I remembered it wrong. I'll rename it. But is this the best way we can deal with such a situation?

Reply:
This is a way I've done clamping in the past. It simply requires a type to be Comparable, so that's pretty broad, and the use of generics prevents type erasure when using the function. Should be pretty good for all our needs.

It's a simple enough function that I might put it in a general HelperFunction.swift or something like that. I imagine there will be a few other functions that might be useful to throw in there.

Reply:
I didn't mean the way to do the clamping.
What I meant was to prevent extremely large values from ruining the plot.

Reply:
That's a good question. clamp's an easy way to handle it at first, given a display range, but you might want to show the discontinuity in the plotted function at the top or bottom of the graph as they extend upward.

Might be worth looking into how other plotting frameworks handle this, because I can't remember how we dealt with things like asymptotes before.

Reply:
Alright, I'll search for it. I plotted 1/x in matplotlib, I didn't see the code, but from the graph it seemed like they were skipping the infinity points and then clamping the large values. I thought this because of a near vertical line at x = 0

Reply:
If the values were clamped to ranges that were just offscreen, you might be able to show the appropriate discontinuities on the edges and potentially draw the lines that connect really large values to really small ones. That could be a simple solution that looks nice onscreen.

Reply:
The problem here would be what value would I clamp it to? I'm just querying for y values for each x. Then I decide the x and y ranges according to maxX maxY minX minY. When I don't clamp and the absolute maxX and minY have a large difference, we won't be able to se either one of them. I think I'll be able to explain the problem better with an image. I'll do that when I get home.

Reply:
**with clamping**
![_06_function_plot_line_chart](https://user-images.githubusercontent.com/35020024/58119123-e9cf0600-7c1f-11e9-9570-862695f2e487.png)
**without clamping**
![_06_function_plot_line_chart](https://user-images.githubusercontent.com/35020024/58119182-066b3e00-7c20-11e9-8b5a-42636d2b2a54.png)

This is 1/x

r -5.000002
r -5.714288
r -6.66667
r -8.000005
r -10.000008
r -13.333346
r -20.000029
**r -40.00012**
**r 13421773.0**
r 39.99988
r 19.99997
r 13.33332
r 9.999992
r 7.999995
r 6.666663
r 5.7142835

These are the values of the function close to 0
max Y = 13421773
min Y = -40
So in comparison to maxY minY is very small.

Reply:
@BradLarson It seems matplotlib has the same problem like above
I tried plotting for different **x** ranges and saw the same problem.
But when I tried for just positive or negative numbers they were clamping at the inverse of the step value.
Like:
```
import matplotlib
import matplotlib.pyplot as plt
import numpy as np

t = np.arange(-2.0, 0.0, 0.001) #here the graph is clamped at 1000 or -1000 whichever is applicable
s = 1/t

fig, ax = plt.subplots()
ax.plot(t, s)

ax.set(xlabel='time (s)', ylabel='voltage (mV)',
       title='About as simple as it gets, folks')
ax.grid()

plt.show()
```

I think we too can clamp at the inverse of the step value.

Reply:
That makes sense. If the user manually fixes a range for the viewport of the graph, you could easily clamp to that, but following matplotlib's behavior is sensible in the case you describe.

Thanks for doing the legwork to research this, I couldn't remember how we handled this case before in Core Plot.